File: DSA/src/Questions/DP/UnboundedKnapsack.java
package Questions.DP;

import java.util.Arrays;

public class UnboundedKnapsack {

    // Recursive function to solve the unbounded knapsack problem
    static int knapsackUtil(int[] wt, int[] val, int ind, int W, int[][] dp) {
        // Base case: If there are no more items to consider
        if (ind == 0) {
            // Calculate and return the maximum value possible
            return ((int) (W / wt[0])) * val[0];
        }

        // If the result for this subproblem has already been calculated, return it
        if (dp[ind][W] != -1)
            return dp[ind][W];

        // Calculate the maximum value when the current item is not taken
        int notTaken = 0 + knapsackUtil(wt, val, ind - 1, W, dp);

        // Initialize the maximum value when the current item is taken as the minimum
        // integer value
        int taken = Integer.MIN_VALUE;

        // If the weight of the current item is less than or equal to the available
        // capacity (W),
        // calculate the maximum value when the current item is taken
        if (wt[ind] <= W)
            taken = val[ind] + knapsackUtil(wt, val, ind, W - wt[ind], dp);

        // Store the result in the dp array and return it
        return dp[ind][W] = Math.max(notTaken, taken);
    }

    // Function to find the maximum value of items that the thief can steal
    static int unboundedKnapsackRecursive(int n, int W, int[] val, int[] wt) {
        // Create a 2D array to store results of subproblem
        int[][] dp = new int[n][W + 1];

        // Initialize the dp array with -1 to indicate that subproblem are not solved
        // yet
        for (int[] row : dp)
            Arrays.fill(row, -1);

        // Call the knapsackUtil function to solve the problem
        return knapsackUtil(wt, val, n - 1, W, dp);
    }
    // Time Complexity: O(N*W)

    // Reason: There are N*W states therefore at max ‘N*W’ new problems will be solved.

    // Space Complexity: O(N*W) + O(N)

    // Function to solve the unbounded knapsack problem
    static int unboundedKnapsackTabulation(int n, int W, int[] val, int[] wt) {
        // Create a 2D array to store results of subproblem
        // dp[i][j] stores the maximum value that can be obtained with a knapsack of
        // capacity j and using the first i items
        int[][] dp = new int[n][W + 1];

        // Base condition: Initialize the dp array for the first item
        for (int i = wt[0]; i <= W; i++) {
            dp[0][i] = ((int) i / wt[0]) * val[0];
        }

        // Fill the dp array using dynamic programming
        for (int ind = 1; ind < n; ind++) {
            for (int cap = 0; cap <= W; cap++) {
                // Calculate the maximum value when the current item is not taken
                int notTaken = 0 + dp[ind - 1][cap];

                // Initialize the maximum value when the current item is taken as the minimum
                // integer value
                int taken = Integer.MIN_VALUE;

                // If the weight of the current item is less than or equal to the current
                // capacity (cap),
                // calculate the maximum value when the current item is taken
                if (wt[ind] <= cap)
                    taken = val[ind] + dp[ind][cap - wt[ind]];

                // Store the result in the dp array
                dp[ind][cap] = Math.max(notTaken, taken);
            }
        }

        return dp[n - 1][W]; // Return the maximum value that can be obtained
    }
    // Time Complexity: O(N*W)

    // Reason: There are two nested loops

    // Space Complexity: O(N*W)

    // Function to solve the unbounded knapsack problem
    static int unboundedKnapsack(int n, int W, int[] val, int[] wt) {
        // Create an array to store the maximum value for each capacity from 0 to W
        int cur[] = new int[W + 1];

        // Base condition: Initialize the cur array for the first item
        for (int i = wt[0]; i <= W; i++) {
            cur[i] = ((int) i / wt[0]) * val[0];
        }

        // Fill the cur array using dynamic programming
        for (int ind = 1; ind < n; ind++) {
            for (int cap = 0; cap <= W; cap++) {
                // Calculate the maximum value when the current item is not taken
                int notTaken = cur[cap];

                // Initialize the maximum value when the current item is taken as the minimum
                // integer value
                int taken = Integer.MIN_VALUE;

                // If the weight of the current item is less than or equal to the current
                // capacity (cap),
                // calculate the maximum value when the current item is taken
                if (wt[ind] <= cap)
                    taken = val[ind] + cur[cap - wt[ind]];

                // Store the result in the cur array
                cur[cap] = Math.max(notTaken, taken);
            }
        }

        return cur[W]; // Return the maximum value that can be obtained with the given capacity W
    }
    // Time Complexity: O(N*W)

    // Reason: There are two nested loops.

    // Space Complexity: O(W)

    public static void main(String[] args) {
        // Input: Number of items and the capacity of the knapsack
        int n = 4, W = 8;

        // Arrays to store the values and weights of the items
        int val[] = { 1, 30, 50, 70 };
        int wt[] = { 1, 5, 3, 4 };

        // Function call to find the maximum value of items that the thief can steal
        System.out.println(unboundedKnapsack(n, W, val, wt)); // Output: 120
    }

}

File: DSA/Python/Questions/DP/unbounded_knapsack.py
"""
# Question: Unbounded Knapsack
# Link: https://leetcode.com/problems/unbounded-knapsack/

# Problem Statement:
# Given a set of N items with certain weight and value, and a knapsack of capacity W.
# Each item can be picked any number of times. Find the maximum value that can be achieved.

# Example:
# Input: N = 4, W = 8
# values = [1, 30, 50, 70]
# weights = [1, 5, 3, 4]
# Output: 120
"""

from typing import List

# 1. Recursive Solution with Memoization
"""
Algorithm:
1. For each item, we have two choices:
   - Take current item (can take multiple times)
   - Skip current item
2. Track remaining capacity and current index
3. Use dp table to store computed results

Time Complexity: O(N*W)
Space Complexity: O(N*W) + O(N) for recursion stack
"""


def unbounded_knapsack_recursive(
    weights: List[int], values: List[int], capacity: int
) -> int:
    n = len(weights)
    dp = [[-1] * (capacity + 1) for _ in range(n)]

    def solve(ind: int, W: int) -> int:
        # Base case: first item
        if ind == 0:
            return (W // weights[0]) * values[0]

        if dp[ind][W] != -1:
            return dp[ind][W]

        # Don't take current item
        not_taken = solve(ind - 1, W)

        # Take current item if possible
        taken = 0
        if weights[ind] <= W:
            taken = values[ind] + solve(ind, W - weights[ind])

        dp[ind][W] = max(not_taken, taken)
        return dp[ind][W]

    return solve(n - 1, capacity)


# 2. Tabulation Solution
"""
Algorithm:
1. Create dp table of size n * (capacity + 1)
2. Initialize first row based on first item
3. For each item and capacity:
   - Calculate value without taking item
   - Calculate value with taking item
   - Store maximum value

Time Complexity: O(N*W)
Space Complexity: O(N*W)
"""


def unbounded_knapsack_tabulation(
    weights: List[int], values: List[int], capacity: int
) -> int:
    n = len(weights)
    dp = [[0] * (capacity + 1) for _ in range(n)]

    # Initialize first row
    for w in range(capacity + 1):
        dp[0][w] = (w // weights[0]) * values[0]

    # Fill dp table
    for ind in range(1, n):
        for cap in range(capacity + 1):
            not_taken = dp[ind - 1][cap]

            taken = 0
            if weights[ind] <= cap:
                taken = values[ind] + dp[ind][cap - weights[ind]]

            dp[ind][cap] = max(not_taken, taken)

    return dp[n - 1][capacity]


# 3. Space Optimized Solution
"""
Algorithm:
1. Use single array instead of 2D array
2. Update array for each item
3. Process from left to right

Time Complexity: O(N*W)
Space Complexity: O(W)
"""


def unbounded_knapsack_optimized(
    weights: List[int], values: List[int], capacity: int
) -> int:
    n = len(weights)
    curr = [0] * (capacity + 1)

    # Initialize array
    for w in range(capacity + 1):
        curr[w] = (w // weights[0]) * values[0]

    # Process each item
    for ind in range(1, n):
        for cap in range(capacity + 1):
            not_taken = curr[cap]

            taken = 0
            if weights[ind] <= cap:
                taken = values[ind] + curr[cap - weights[ind]]

            curr[cap] = max(not_taken, taken)

    return curr[capacity]


def main():
    test_cases = [
        {"weights": [1, 5, 3, 4], "values": [1, 30, 50, 70], "capacity": 8},
        {"weights": [2, 4, 6], "values": [5, 11, 13], "capacity": 10},
    ]

    for i, test in enumerate(test_cases):
        print(f"\nTest Case {i + 1}:")
        print(f"Weights: {test['weights']}")
        print(f"Values: {test['values']}")
        print(f"Capacity: {test['capacity']}")
        print(
            f"Recursive Solution: {unbounded_knapsack_recursive(test['weights'], test['values'], test['capacity'])}"
        )
        print(
            f"Tabulation Solution: {unbounded_knapsack_tabulation(test['weights'], test['values'], test['capacity'])}"
        )
        print(
            f"Optimized Solution: {unbounded_knapsack_optimized(test['weights'], test['values'], test['capacity'])}"
        )


if __name__ == "__main__":
    main()
